plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -0,0 +1,618 @@
+				     guint8 target_addr, guint8 sub_addr,
+				     guint8 *data, gsize datasz,
+				     GError **error)
+{
+	FuRts54HubDevice *parent;
+
+	parent = fu_rts54hub_rtd21xx_device_get_parent (self, error);
+	if (parent == NULL)
+		return FALSE;
+	if (!fu_rts54hub_device_vendor_cmd (parent, 1, error))
+		return FALSE;
+	if (target_addr != self->target_addr) {
+		if (!fu_rts54hub_device_i2c_config (parent, target_addr, 1,
+						    FU_RTS54HUB_I2C_SPEED_200K,
+						    error))
+			return FALSE;
+		self->target_addr = target_addr;
+	}
+	if (!fu_rts54hub_device_i2c_read (parent, sub_addr, data, datasz, error)) {
+		g_prefix_error (error, "failed to read I2C: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_read_status_raw (FuRts54hubRtd21xxDevice *self,
+					    guint8 *status,
+					    GError **error)
+{
+	guint8 buf[] = { 0x00 };
+	if (!fu_rts54hub_rtd21xx_device_i2c_read (self,
+						  UC_FOREGROUND_SLAVE_ADDR,
+						  UC_FOREGROUND_STATUS,
+						  buf, sizeof(buf),
+						  error))
+		return FALSE;
+	if (status != NULL)
+		*status = buf[0];
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_read_status_cb (FuDevice *device,
+					   gpointer user_data,
+					   GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	guint8 status = 0xfd;
+	if (!fu_rts54hub_rtd21xx_device_read_status_raw (self, &status, error))
+		return FALSE;
+	if (status == ISP_STATUS_BUSY) {
+		g_set_error (error,
+				FWUPD_ERROR,
+				FWUPD_ERROR_INTERNAL,
+				"status was 0x%02x", status);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_read_status (FuRts54hubRtd21xxDevice *self,
+					guint8 *status,
+					GError **error)
+{
+	return fu_device_retry (FU_DEVICE (self),
+				fu_rts54hub_rtd21xx_device_read_status_cb,
+				4200, status, error);
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_ensure_version_unlocked (FuRts54hubRtd21xxDevice *self,
+					     GError **error)
+{
+	guint8 buf_rep[7] = { 0x00 };
+	guint8 buf_req[] = { ISP_CMD_GET_FW_INFO };
+	g_autofree gchar *version = NULL;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_FOREGROUND_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   buf_req, sizeof(buf_req),
+						   error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+
+	/* wait for device ready */
+	g_usleep(300000);
+	if (!fu_rts54hub_rtd21xx_device_i2c_read (self,
+						  UC_FOREGROUND_SLAVE_ADDR,
+						  0x00,
+						  buf_rep, sizeof(buf_rep),
+						  error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+	/* set version */
+	version = g_strdup_printf ("%u.%u", buf_rep[1], buf_rep[2]);
+	fu_device_set_version (FU_DEVICE (self), version);
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_detach_raw (FuRts54hubRtd21xxDevice *self, GError **error)
+{
+	guint8 buf[] = { 0x03 };
```

Comment:
can this be changed into `guint8 buf = 0x03;` and then use derferencing operator in the function call instead?

Reply:
can indeed, fixed.